File: JavaScript/Queue/1-linkedlist.js
/**
 * My implementation of the "Queue"
 */

'use strict';

class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.root = null;
    this.tail = null;
  }

  // push
  enqueue(data) {
    const node = new Node(data);
    if (this.root === null) {
      this.root = node;
      this.tail = node;
      return this;
    }
    this.tail.next = node;
    this.tail = node;
    return this;
  }

  // shift
  dequeue() {
    if (!this.root && !this.tail) {
      throw new Error('Nothing is in the queue');
    }
    if (this.root === this.tail) {
      this.root = null;
      this.tail = null;
      return this;
    }
    this.root = this.root.next;
    return this;
  }
}

// usage

const queue = new Queue();

queue.enqueue(1); // root
queue.enqueue(2);
queue.enqueue(3);
queue.enqueue(4); // tail
console.dir({ queue }, { depth: null });

queue.dequeue(); // 2 - root, 4 - tail
console.dir({ queue }, { depth: null });

File: C/Queue/1-linkedlist.c
#include <stdio.h> 
#include <stdlib.h> 

// Node structure
struct Node {
    int data;
    struct Node* next; // pointer to next element
};

struct Node* head = NULL;
struct Node* tail = NULL;

// push to the end
void Enqueue(int data) {
    // creating new Node, allocating memory for structure
    struct Node* temp = 
        (struct Node*)malloc(sizeof(struct Node));
    temp->data = data; // Node's data
    temp->next = NULL; // next element will be undefined, so this Node is last in the queue
    // if there is no elements yet, create one
    if (head == NULL && tail == NULL) {
        head = temp;
        tail = temp;
        return;
    }
    // if we do have elements, then just add new one to the end
    tail->next = temp; // last element points to new node
    tail = temp; // real address of tail changes to address of new node
    return;
}

void Dequeue() {
    struct Node* temp = head; // copying head
    if (head == NULL) return; // if there is no element 
    if (head == tail) { // if there is only one element
        head = NULL;
        tail = NULL;
    }
    else {
        head = head->next; // moving head to the next element
    }
    free(temp); // completely removing Node from memory
    return;
}

void PrintQueue() {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}


int main() {
    Enqueue(1); PrintQueue(); // 1
    Enqueue(2); PrintQueue(); // 1 2
    Enqueue(3); PrintQueue(); // 1 2 3
    Dequeue();  PrintQueue(); // 2 3
    Enqueue(4); PrintQueue(); // 2 3 4
    
    return 0;
}